keps/sig-apps/4563-eviction-request-api/README.md

```diff
@@ -0,0 +1,1948 @@
+owner/controller of the pod is the main interceptor. In a special case, the pod can be its own
+interceptor. The interceptor should decide what action to take when it observes an eviction request
+intent directed at that interceptor:
+1. It can reject the eviction request and wait for the pod to be intercepted by another interceptor
+   or evicted by the eviction request controller.
+2. It can do nothing and wait for the pod to be intercepted by another interceptor or evicted by the
+   eviction request controller. This is discouraged because it is slower than a simple rejection.
+3. It can start the eviction logic and periodically respond to the eviction request intent to signal
+   that the eviction request is in progress and not stuck. The eviction logic is at the discretion
+   of the interceptor and can take many forms:
+   - Migration of data (both persistent and ephemeral) from one node to another.
+   - Waiting for a cleanup and release of important resources held by the pod.
+   - Waiting for important computations to complete.
+   - Non-graceful termination of the pod (`gracePeriodSeconds=0`).
+   - Deletion of a pod that is covered by a blocking PodDisruptionBudget. The controller of the
+     application should have additional logic to distinguish whether a disruption of a particular
+     pod will disrupt the application as a whole.
+   - After a partial cleanup (e.g. storage migrated, notification sent) and if the application is
+     still in an available state, the API-initiated eviction can be used to respect
+     PodDisruptionBudgets.
+
+The end goal of an EvictionRequest is for a pod to be terminated by one of the interceptors, or by
+an API-initiated eviction triggered by the eviction request controller. Usually this will also
+coincide with the deletion of the pod (evict or delete call). In some scenarios, the pod may only be
+terminated (e.g. by a remote call) if the pod `restartPolicy` allows it, to preserve the pod data
+for further processing or debugging.
+
+The interceptor can also choose whether it handles EvictionRequest cancellation. See
+[EvictionRequest Cancellation Examples](#evictionrequest-cancellation-examples) for details.
+
+We should discourage the creation of preventive EvictionRequests, so that they do not end up as
+another PDB. So we should design the API appropriately and also not allow behaviors that do not
+conform to the eviction request contract.
+
+### Eviction Request Controller
+
+In order to fully enforce the eviction request contract and prevent code duplication among eviction
+requesters, we will introduce a new controller called the eviction request controller.
+
+Its responsibility is to observe eviction requests from requesters and periodically check that
+interceptors are making progress in evicting/terminating pods. It is important to see a consistent
+effort by the interceptors to reconcile the progress of the eviction request. This is important to
```

Comment:
We do not need to enforce the progress. This is modelled after leader election. If an interceptor fails to update the EvictionRequest status to report the progress we will choose another interceptor or evict the pod.

Leader election works fine with much lower timeouts. I think we can be more forgiving here, but the 10m mark (allowed minimum) should be sufficient.

> that can be overused by dummy updates, just to block longer/indefinitely.

The same could be said about the leader election, which could be used to hijack an existing controller.

The idea behind this is that we should make it very unpleasant to block eviction  (proof of work instead of just creating a PDB). And if someone implements this, they might as well implement graceful termination.

Another benefit of this approach is that an administrator can take a look at a very long EvictionRequest and see what InterceptorClass is set on it and can easily track down the misbehaving actor. We can also hook into this with metrics/alerts.

Finally, since this is a normal k8s object, we can use RBAC to restrict and observer access to the EvictionRequest.